### PR TITLE
[flang][OpenMP] Normalize clause modifiers that exist on their own

### PR DESCRIPTION
**Pull request by kparzysz**
This is the first part of the effort to make parsing of clause modifiers more uniform and robust. Currently, when multiple modifiers are allowed, the parser will expect them to appear in a hard-coded order. Additionally, modifier properties (such as "ultimate") are checked separately for each case.

The overall plan is
1. Extract all modifiers into their own top-level classes, and then equip them with sets of common properties that will allow performing the property checks generically, without refering to the specific kind of the modifier.
2. Define a parser (as a separate class) for each modifier.
3. For each clause define a union (std::variant) of all allowable modifiers, and parse the modifiers as a list of these unions.

The intent is also to isolate parts of the code that could eventually be auto-generated.

OpenMP modifier overhaul: #1/3

### CODE REVIEW

**Review by kiranchandramohan — COMMENTED** [review]
Is this only for modifiers that can appear together or for all modifiers. What about ScheduleModifier?

**Comment by kparzysz** [comment]
Eventually this is for all modifiers, but I want to do this in smaller steps.  One extra goal I'm trying to accomplish is to make it easy to add the new `directive-name-modifier`.

At the end of all this, every modifier will be a top-level class (i.e. not defined inside of a clause).  Then every clause will use the "variant" instead of putting modifiers explicitly in a tuple.

File: flang/examples/FlangOmpReport/FlangOmpReportVisitor.cpp [hunk 1]
```diff
@@ -218,11 +218,11 @@ void OpenMPCounterVisitor::Post(const OmpScheduleModifierType::ModType &c) {
   clauseDetails +=
```
[the remark this answers is not in the document]
Comment (kparzysz) [reply flang/examples/FlangOmpReport/FlangOmpReportVisitor.cpp:218]: I renamed the `OmpReductionOperator` to `OmpReductionIdentifier` to follow the naming from the spec.

**Comment by kparzysz** [comment]
I'm actually working on the `schedule` clause now.  Almost ready.

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Review by tblah — APPROVED** [review]
Thanks for the cleanup!

**Pull request merged**